[rustpotterks] Upgrade to version 3

GiviMAD opened the pull request:
I would like to merge this upgrade to the rustpotterks binding which uses the latest version of the library.

This version of the library, probably the final one, introduces a new "wakeword" type that relies on training a classification network to do the detections. I have been some weeks using it and work by far better than the previous versions.
The files from the version 2 are still compatibles, called wakeword references in this version, to differentiate them from the trained models. But I would recommend not using them, because the version 2 of the library uses frames of 5 mfccs, and that number seems to be too low, that parameter is now configurable when generating the wakewords and defaults to a more appropriate value, 16.

I've introduced a change in the addon to update the configuration of the running instances, so you don't need to reset your dialog in order to get the new configuration, and also I have marked as advanced all the options that are not required in order to use it, I think it looks more user friendly like this. 

Apart from that, I've discovered a bug. I faced it some weeks ago but I wasn't able to discover the cause, but when implementing the configuration update as an asynchronous task I realize that the use of the ScheduledExecutorService causes the task to not run in parallel, and the configuration update was delayed until I have stopped the dialog execution. It also causes that several instances can not run in parallel with is the main purpose of the service. I think I didn't realize of this in the v3 addon version because there we have a `Thead.sleep` call inside the loop which was called eventually. After changing it to an ExecutorService it run without problems, the configuration updates are correctly propagated and I can run multiple dialog instances using rustpotter without breaking openHAB, which is what happened to me a couple of weeks ago, but as I was testing the habspeaker addon I assumed it was due to some error there, but this seems to be the cause.
I think that maybe the executor type should be changed in other voice add-ons to prevent these problem, as those are also intended to run in parallel, maybe @wborn you can give some guidance on this? I'm a little lost with all the different executor classes.

wborn commented:
> I think that maybe the executor type should be changed in other voice add-ons to prevent these problem, as those are also intended to run in parallel

Why not just increase the threads in the pool to allow for more threads to run in parallel?

GiviMAD commented:
> Why not just increase the threads in the pool to allow for more threads to run in parallel?

Thank for the reply, I didn't realize at that moment that the ThreadPoolManager was an OpenHAB core util to allow configuring the pool size though the application, told you I was a little lost :) . I have reverted back using it but changed it to from a scheduled pool to a normal one also I have added back the sleep call but with a more appropriate time value.
The thing is that today I'm not able to replicate the issue, maybe there was another thing keeping the threads busy.

wborn commented:
> The thing is that today I'm not able to replicate the issue, maybe there was another thing keeping the threads busy

It could also be your IDE pausing threads when they run into a break point. :wink:

GiviMAD commented:
> It could also be your IDE pausing threads when they run into a break point. 😉

That could be, but no, it was on my "producction" openHAB installation, I have the PR add-on sideloaded there.

But I have been running a persistent dialog execution for several weeks (maybe a month IDK) using the pulseaudio binding, with the whisper addon from the other PR I have opened as STT service, and also using the habspeaker add on from the marketplace a lot testing the integration of rustpotter there, so a lot of untested/lowtested stuff. I'll keep and eye on the thread number to see if I'm able to see this happening again, I'm going to keep testing all of this. 

Thank you again for the support :)

GiviMAD commented:
@wborn finally got it, it was caused by bug in habspeaker, I was incorrectly disposing the pipe that is used to share the audio source data among the different consumers, generating locked threads on the keyword spotter service (rustpotter in this case), at the moment I created 4 locks I get back to the behavior that I saw when I asked for help.

I think I have finally fixed it after refactoring that part of the code a couple of times, I made it a little over complicated the first time, now I have centralized that logic inside the AudioStream implementation which seems to be a better idea. The thread command was really helpful to understand the locks, I think I had already sawn it when I started digging into the project but haven't get advantage of it until this time, it has saved my day :) .

If you have any time to look at this pr, I hope it can get it into the 4.1.0 release it offers a huge improvement over the current version, for me it seems to be prepared to be used on a "production" system. What is it still not working in an acceptable way is the TTS (vosk or whisper), but I still need to learn more about the whisper configuration and test more models, but currently I'm trying to get to a version of habspeaker that can be considered a RC, I have less free time lately due to the work, so I'm trying to close things one after another so the time I can dedicate to them is a little more efficient.

 I will try to train a share a wakeword that is not tied to a language so people can give it a try without having to spend an hour on training a wakeword with the cli, because for developing/testing I used a Spanish phrase. I was thinking on trying to create one for sharing, mixing records of my voice with others generated using a Text to Speech engine, to see how good it goes.

wborn commented on bundles/org.openhab.voice.rustpotterks/README.md:
```diff
@@ -3,12 +3,14 @@
-Rustpotter provides personal on-device wake word detection. You need to generate a model for your keyword using audio samples.
+Rustpotter provides personal on-device wake word detection. You need to generate a file for your keyword using audio samples.
 
 You can test library in your browser using these web pages:
```
```suggestion
You can test the library in your browser using these web pages:
```

wborn commented on bundles/org.openhab.voice.rustpotterks/README.md:
```diff
@@ -3,12 +3,14 @@
 [Rustpotter](https://github.com/GiviMAD/rustpotter) is a free and open-source keywords spotter written in rust.
 
-Rustpotter provides personal on-device wake word detection. You need to generate a model for your keyword using audio samples.
+Rustpotter provides personal on-device wake word detection. You need to generate a file for your keyword using audio samples.
```
```suggestion
Rustpotter provides personal on-device wake word detection.
You need to generate a file for your keyword using audio samples.
```

wborn commented on bundles/org.openhab.voice.rustpotterks/src/main/java/org/openhab/voice/rustpotterks/internal/RustpotterKSService.java:
```diff
@@ -196,10 +223,18 @@ private void processAudioStream(Rustpotter rustpotter, KSListener ksListener, Au
                     remaining = remaining - numBytesRead;
+                    try {
+                        Thread.sleep(remaining / bytesPerMs);
+                    } catch (InterruptedException ignored) {
```
Are you sure it shouldn't be ignored? E.g. when someone shuts down the runtime the thread stop running ASAP.

GiviMAD replied:
Yes, I think you are right, I have added a warning to be aware when these happens and set the aborted bool to true.

wborn commented on bundles/org.openhab.voice.rustpotterks/src/main/resources/OH-INF/i18n/rustpotterks_it.properties:
This file should be updated using Crowdin so it stays in sync.

GiviMAD replied:
Should I do that? I always thought that was done automatically, probably the reviewers were doing it for me.

I've never catch up with the Crowdin integration but it's something I want to explore to see if I can help with Spanish or Galician support.

wborn replied:
It will only upload the default translations to Crowdin. So if you rename keys only on GitHub they keep showing as untranslated on Crowdin.

It's up to you if you also want to update it in Crowdin. I usually do it because it is not a lot of work and Crowdin will usually suggest the previous translation anyways because the text is a 100% match with the previous value in the translation cache.

GiviMAD replied:
I copied the full content of the main file because a many options changed (I think also the descriptions where improved) when I updated the library to the v2 and I totally forgot to do this back then. Hope it's ok, and thank you for the explanation.

wborn commented on bundles/org.openhab.voice.rustpotterks/src/main/java/org/openhab/voice/rustpotterks/internal/RustpotterKSService.java:
```diff
@@ -62,27 +66,41 @@
+    private static final String RUSTPOTTER_RECORDS_FOLDER = Path.of(RUSTPOTTER_FOLDER, "records").toString();
     private final Logger logger = LoggerFactory.getLogger(RustpotterKSService.class);
-    private final ScheduledExecutorService executor = ThreadPoolManager.getScheduledPool("OH-voice-rustpotterks");
+    private final ExecutorService executor = ThreadPoolManager.getPool("voice-rustpotterks");
```
Wil the threads still get an `OH-` prefix?

GiviMAD replied:
That seems to be the case, it was displayed as "OH-OH-voice-rustpotterks" in the output of the `threads` console command.

wborn commented on bundles/org.openhab.voice.rustpotterks/src/main/java/org/openhab/voice/rustpotterks/internal/RustpotterKSService.java:
```diff
@@ -62,27 +66,41 @@
+    }
+
+    private static void tryCreateDir(String rustpotterFolder, Logger logger, String msg) {
+        File addonDir = new File(rustpotterFolder);
+        if (!addonDir.exists()) {
+            if (addonDir.mkdir()) {
+                logger.info(msg, rustpotterFolder);
             }
         }
```
As you already use `java.nio.file.Path` you might as well stop using `java.io.File` here as well and replace it with `Files.createDirectories`. In general using `java.nio` is preferred over `java.io.File` because it has nicer APIs and better performance.

GiviMAD replied:
Thank you for the suggestion, I wasn't aware of that, I have moved the code to use Path instances and the helpers of the Files class, and as you said it looks much better.

One side question about this, more related to the service design. Can I move the folder creation from the static context to the activate method and throw the exceptions to prevent the add-on from been loaded if the folder creation doesn't succeed? I don't know if is correct to throw exceptions from the activate method or if those should be catch inside it.

wborn replied:
Creating the folders in the activate method is probably better so they will be recreated (when missing) every time the bundle is restarted. If thrown exceptions result in big stacktraces it might be better to catch exceptions and log a helpful message instead.

GiviMAD commented on bundles/org.openhab.voice.rustpotterks/src/main/java/org/openhab/voice/rustpotterks/internal/RustpotterKSService.java:
```diff
@@ -154,40 +175,46 @@ public KSServiceHandle spot(KSListener ksListener, AudioStream audioStream, Loca
 
-    private void processAudioStream(Rustpotter rustpotter, KSListener ksListener, AudioStream audioStream,
-            AtomicBoolean aborted) {
+    private void setBindingOptions(RustpotterKSConfiguration bindingConfig, RustpotterConfig rustpotterConfig) {
```
Yes, it was a pretty bad naming choice from my side, and it doesn't need to be like that at all, thank you for pointing at it.

I have renamed it to initRustpotterConfig which returns a new library config instance already updated with the service configuration.

wborn reviewed (approved):
Thanks!

GiviMAD commented:
Thank you!
